### PR TITLE
Fix calling convention mismatch in GC callouts

### LINKED ISSUE #110607: GC crash in .NET 9 UWP and WinUI 3 apps published with Native AOT on win-x86
### Description

I am migrating an UWP applicaiton to .NET 9 and found a problem with win-x86 Native AOT publish. The app works when it's either unpackaged or published using win-x64 runtime. But when it's published using win-x86 runtime, gc crashes on random instructions when trying to collect memory. Example stack trace from crash dump:
```
The thread tried to read from or write to a virtual address for which it does not have the appropriate access.

UwpDotnet.exe!WKS::gc_heap::mark_phase(int condemned_gen_number=5150738)	C++
UwpDotnet.exe!WKS::gc_heap::gc1()	C++
UwpDotnet.exe!WKS::gc_heap::garbage_collect(int n)	C++
UwpDotnet.exe!WKS::GCHeap::GarbageCollectGeneration(unsigned int gen=0, gc_reason reason=reason_alloc_soh)	C++
UwpDotnet.exe!WKS::gc_heap::trigger_gc_for_alloc(int gen_number=0, gc_reason gr=reason_alloc_soh, WKS::GCDebugSpinLock * msl=0x0082f010, bool loh_p=false, WKS::msl_take_state take_state=mt_try_budget)	C++
[Inline Frame] UwpDotnet.exe!WKS::gc_heap::try_allocate_more_space(alloc_context *)	C++
UwpDotnet.exe!WKS::gc_heap::allocate_more_space(alloc_context * acontext=0x01640358, unsigned int size=4012, unsigned int flags=0, int alloc_generation_number=0)	C++
[Inline Frame] UwpDotnet.exe!WKS::gc_heap::allocate(unsigned int)	C++
UwpDotnet.exe!WKS::GCHeap::Alloc(gc_alloc_context * context=0x0082f010, unsigned int size=4012, unsigned int flags=0)	C++
UwpDotnet.exe!GcAllocInternal(MethodTable * pEEType=0x008f319c, unsigned int uFlags=0, unsigned int numElements=1000, Thread * pThread=0x01640358)	C++
UwpDotnet.exe!RhpGcAlloc(MethodTable * pEEType=0x008f319c, unsigned int uFlags=0, unsigned int numElements=1000, PInvokeTransitionFrame * pTransitionFrame=0x0782f8dc)	C++
UwpDotnet.exe!@RhpNewArray@8()	Unknown
>	UwpDotnet.exe!UwpDotnet_UwpDotnet_Program__RunLoad()	Unknown
UwpDotnet.exe!S_P_CoreLib_System_Threading_ThreadStart__InvokeOpenStaticThunk()	Unknown
UwpDotnet.exe!S_P_CoreLib_System_Threading_Thread__StartThread()	Unknown
UwpDotnet.exe!S_P_CoreLib_System_Threading_Thread__ThreadEntryPoint()	Unknown
ntdll.dll!77cc809e()	Unknown
[Frames below may be incorrect and/or missing, no symbols loaded for ntdll.dll]	
ntdll.dll!77cc806e()	Unknown
```

The same problem is reproduced for both WinUI 3 and UWP apps, but not reproduces for console application.

### Reproduction Steps

It's reproducible on both UWP and WinUI 3, but for simplicity specifying WinUI 3 here.

1. Create WinUI 3 app on _**net9.0-windows10.0.26100.0**_
```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <OutputType>WinExe</OutputType>
    <TargetFramework>net9.0-windows10.0.26100.0</TargetFramework>
    <TargetPlatformMinVersion>10.0.17763.0</TargetPlatformMinVersion>
    <RootNamespace>WinUIAppTest</RootNamespace>
    <ApplicationManifest>app.manifest</ApplicationManifest>
    <Platforms>x86;x64;arm64</Platforms>
    <RuntimeIdentifiers>win-x86;win-x64;win-arm64</RuntimeIdentifiers>
    <PublishProfile>win-$(Platform).pubxml</PublishProfile>
    <UseWinUI>true</UseWinUI>
    <EnableMsixTooling>true</EnableMsixTooling>
    <Nullable>enable</Nullable>
    <AppendRuntimeIdentifierToOutputPath>false</AppendRuntimeIdentifierToOutputPath>
    <AppendTargetFrameworkToOutputPath>false</AppendTargetFrameworkToOutputPath>
    <PackageCertificateKeyFile>../TemporaryKey.pfx</PackageCertificateKeyFile>
  </PropertyGroup>

  <ItemGroup>
    <Content Include="Assets\SplashScreen.scale-200.png" />
    <Content Include="Assets\LockScreenLogo.scale-200.png" />
    <Content Include="Assets\Square150x150Logo.scale-200.png" />
    <Content Include="Assets\Square44x44Logo.scale-200.png" />
    <Content Include="Assets\Square44x44Logo.targetsize-24_altform-unplated.png" />
    <Content Include="Assets\StoreLogo.png" />
    <Content Include="Assets\Wide310x150Logo.scale-200.png" />
  </ItemGroup>

  <ItemGroup>
    <Manifest Include="$(ApplicationManifest)" />
  </ItemGroup>

  <ItemGroup Condition="'$(DisableMsixProjectCapabilityAddedByProject)'!='true' and '$(EnableMsixTooling)'=='true'">
    <ProjectCapability Include="Msix" />
  </ItemGroup>
  <ItemGroup>
    <PackageReference Include="Microsoft.Windows.SDK.BuildTools" Version="10.0.26100.1742" />
    <PackageReference Include="Microsoft.WindowsAppSDK" Version="1.6.240923002" />
  </ItemGroup>

  <PropertyGroup Condition="'$(DisableHasPackageAndPublishMenuAddedByProject)'!='true' and '$(EnableMsixTooling)'=='true'">
    <HasPackageAndPublishMenu>true</HasPackageAndPublishMenu>
  </PropertyGroup>
</Project>
```
2.  Configure NativeAOT publish in _win-x86.pubxml_
```xml
<?xml version="1.0" encoding="utf-8"?>
<Project ToolsVersion="4.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
  <PropertyGroup>
    <PublishProtocol>FileSystem</PublishProtocol>
    <Platform>x86</Platform>
    <RuntimeIdentifier>win-x86</RuntimeIdentifier>
    <PublishAot>true</PublishAot>
    <SelfContained>true</SelfContained>
  </PropertyGroup>
</Project>
```
3. Add some GC-intensive worker in _App.xaml.cs_
```cs
using System;
using System.Collections.Concurrent;
using System.Threading;
using Microsoft.UI.Xaml;

namespace WinUIAppTest;

public partial class App : Application
{
    private Window? m_window;

    public App()
    {
        InitializeComponent();

        var t = new Thread(RunLoad);
        t.Start();
    }

    protected override void OnLaunched(LaunchActivatedEventArgs args)
    {
        m_window = new MainWindow();
        m_window.Activate();
    }

    public static void RunLoad()
    {
        var queue = new ConcurrentQueue<object>();
        while (true)
        {
            var obj = new
            {
                Property1 = new int[1000]
            };
            if (queue.Count > 100000)
            {
                queue.TryDequeue(out _);
            }

            queue.Enqueue(obj);
            if (Random.Shared.Next() % 1000 == 0)
            {
                Thread.Sleep(10);
            }
        }
    }
}
```
4. Publish app
```ps
dotnet publish /p:Platform=x86 /p:GenerateAppxPackageOnBuild=true
```
5. Install msix package and run app

### Expected behavior

App launches and runs.

### Actual behavior

App crashes with stacktrace specified in descripiton.

### Regression?

No, since x86 is available in .NET 9 only.

### Known Workarounds

_No response_

### Configuration

Dotnet info:
```
.NET SDK:
 Version:           9.0.100
 Commit:            59db016f11
 Workload version:  9.0.100-manifests.4a280210
 MSBuild version:   17.12.7+5b8665660

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19045
 OS Platform: Windows
 RID:         win-x64
 Base Path:   C:\Program Files\dotnet\sdk\9.0.100\

.NET workloads installed:
 [android]
   Installation Source: SDK 9.0.100, VS 17.12.35521.163
   Manifest Version:    35.0.7/9.0.100
   Manifest Path:       C:\Program Files\dotnet\sdk-manifests\9.0.100\microsoft.net.sdk.android\35.0.7\WorkloadManifest.json
   Install Type:              Msi

 [aspire]
   Installation Source: SDK 9.0.100
   Manifest Version:    8.2.2/8.0.100
   Manifest Path:       C:\Program Files\dotnet\sdk-manifests\8.0.100\microsoft.net.sdk.aspire\8.2.2\WorkloadManifest.json
   Install Type:              Msi

 [ios]
   Installation Source: SDK 9.0.100, VS 17.12.35521.163
   Manifest Version:    18.1.9163/9.0.100
   Manifest Path:       C:\Program Files\dotnet\sdk-manifests\9.0.100\microsoft.net.sdk.ios\18.1.9163\WorkloadManifest.json
   Install Type:              Msi

 [maccatalyst]
   Installation Source: SDK 9.0.100, VS 17.12.35521.163
   Manifest Version:    18.1.9163/9.0.100
   Manifest Path:       C:\Program Files\dotnet\sdk-manifests\9.0.100\microsoft.net.sdk.maccatalyst\18.1.9163\WorkloadManifest.json
   Install Type:              Msi

 [macos]
   Installation Source: SDK 9.0.100
   Manifest Version:    15.1.9163/9.0.100
   Manifest Path:       C:\Program Files\dotnet\sdk-manifests\9.0.100\microsoft.net.sdk.macos\15.1.9163\WorkloadManifest.json
   Install Type:              Msi

 [maui-windows]
   Installation Source: SDK 9.0.100, VS 17.12.35521.163
   Manifest Version:    9.0.0/9.0.100
   Manifest Path:       C:\Program Files\dotnet\sdk-manifests\9.0.100\microsoft.net.sdk.maui\9.0.0\WorkloadManifest.json
   Install Type:              Msi

Configured to use loose manifests when installing new manifests.

Host:
  Version:      9.0.0
  Architecture: x64
  Commit:       9d5a6a9aa4

.NET SDKs installed:
  7.0.410 [C:\Program Files\dotnet\sdk]
  9.0.100 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 6.0.36 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.20 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 8.0.11 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 9.0.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 6.0.36 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.20 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 8.0.11 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 9.0.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 6.0.36 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 7.0.20 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 8.0.11 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 9.0.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

Other architectures found:
  x86   [C:\Program Files (x86)\dotnet]
    registered at [HKLM\SOFTWARE\dotnet\Setup\InstalledVersions\x86\InstallLocation]

Environment variables:
  Not set

global.json file:
  Not found
```

Windows SDKs installed;
- 10.0.26100.0
- 10.0.22621.0
- 10.0.18362.0
- 10.0.17763.0
- 10.0.16299.0

### Other information

Repo with reproduction examples: https://github.com/andrew-kulikov/native-aot-test

### PR DESCRIPTION
**Pull request by MichalStrehovsky**
Fixes #110607.

The native side expected fastcall, managed side wasn't fastcall.

Filed #110684 on the test hole. We would have caught it during native AOT x86 bringup if this had any tests since this is a guaranteed stack corruption and crash.

Cc @dotnet/interop-contrib

### CODE REVIEW

**Review by AaronRobinsonMSFT — APPROVED** [review]

**Review by jkoritzinsky — CHANGES_REQUESTED** [review]
Does the JIT support fastcall here? I'm pretty sure this will just fail with another exception. I don't think we ever brought Fastcall support online.

Can we add a test or run some existing test suite on the PR?

**Review by jkoritzinsky — APPROVED** [review]
Okay I went looking through the code and found something very interesting. It looks like when I did the unmanaged calling-convention support rewrite in the JIT, I implemented fastcall correctly.

However, as we have never supported it for P/Invokes, our exceptions that block fastcall are P/Invoke-only. Additionally, the exceptions are only thrown in CoreCLR (NativeAOT likely fails in a more inscrutible way).

In any case, looks like `Fastcall` + `UnmanagedCallersOnly` should work here just fine.

File: src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/TrackerObjectManager.NativeAot.cs
Comment (jkotas) [inline src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/TrackerObjectManager.NativeAot.cs:140]: Can we change the unmanaged side to use the default calling convention instead?

x86 custom calling conventions are pain. It is best to stick to the default calling convention where possible.

Comment (jkotas) [reply src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/TrackerObjectManager.NativeAot.cs:140]: We have two different callouts https://`github.com/dotnet/runtime/blob/90dfc04d8f863e053e42edf0992c1b63398b12fb/src/coreclr/nativeaot/Runtime/RestrictedCallouts.h#L100-L101 . One of them is FastCall and the other one is default. I would make both of them default.

**Comment by MichalStrehovsky** [comment]
> Does the JIT support fastcall here? I'm pretty sure this will just fail with another exception. I don't think we ever brought Fastcall support online.

I did test this and it worked.

Looking through our test tree, I don't see any test exercising Fastcall though. Do we have any testing for this? Should we have testing? It doesn't look like CoreCLR VM blocks UnmanagedCallersOnly with fastcall either.

Either way, I made this use not-fastcall per Jan's request.

**Comment by am11** [comment]
FCall callbacks probably are implicitly tested assuming tests exercise this code:

```
src/coreclr/nativeaot/Runtime/interoplibinterface.h:    using TryGetCallback = void*(F_CALL_CONV *)(void);
src/coreclr/nativeaot/Runtime/interoplibinterface.h:    using TryGetTaggedMemoryCallback = CLR_BOOL(F_CALL_CONV *)(_In_ Object *, _Out_ void **);
src/coreclr/nativeaot/Runtime/interoplibinterface.h:    using BeginEndCallback = void(F_CALL_CONV *)(void);
src/coreclr/nativeaot/Runtime/interoplibinterface.h:    using IsReferencedCallback = int(F_CALL_CONV *)(_In_ void*);
src/coreclr/nativeaot/Runtime/interoplibinterface.h:    using EnteredFinalizationCallback = void(F_CALL_CONV *)(_In_ void*);
```
(math functions are also using F_CALL_CONV on x86/arm32 but they are not callbacks)

**Comment by MichalStrehovsky** [comment]
> FCall callbacks probably are implicitly tested assuming tests exercise this code:

Do any of those call into managed code? We're not discussing test coverage of fastcall in the Visual C++ compiler but `UnmanagedCallersOnly` methods with `CallConvs = typeof(CallConvFastcall)`.

**Review by jkotas — APPROVED** [review]

**Comment by MichalStrehovsky** [comment]
This is fixed in main. I consider backport to 9.0 blocked on #110684 - I don't want to be laughed out of the shiproom when I say "no we don't have test coverage, but we do have a concept of a plan for it".

**Comment by Sergio0694** [comment]
Would it be sufficient to validate this on both x86 and x64 with the Store? We won't be able to ship to x86 with Native AOT otherwise. In theory we could keep that target on .NET Native, but it'd be simpler for us if we could migrate all 3 targets. And would the fact x86 right now is flat out broken on NAOT in XAML scenario help make the case for servicing?

**Comment by jkotas** [comment]
If you can validate the fix with Store, it should be good enough test coverage for servicing backport.

**Comment by Sergio0694** [comment]
I tested this on x64 and it works fine (except for #110823, but that's unrelated). Currently blocked by other dependencies on testing with x86, but I hope to get that resolved by end of this week, hopefully. Will report back as soon as I can also validate on x86 and confirm that things work correctly, and then hopefully we can backport this too 🙂

**Comment by MichalStrehovsky** [comment]
> I tested this on x64 and it works fine (except for #110823, but that's unrelated).

The bug can only be observed on x86. There's no fastcall or stdcall on x64, there's only one calling convention. This PR is essentially no-op on x64 or arm64.

**Comment by Sergio0694** [comment]
Oh I know, but I figured it wouldn't hurt to validate everything end to end too just in case, to make things a bit simpler for tactics to backport this. Also until we unblock x86 in the Store it was the only thing I could do 😅

**Comment by Sergio0694** [comment]
Didn't want to delay this any further, so I ended up just yeeting all problematic code from the Store for testing 😆

Followed similar steps as https://github.com/dotnet/runtime/pull/110762#pullrequestreview-2508994178:
Tested with the Microsoft Store and could not repro any hangs:
- Checked out `release/9.0-staging` (f58036422045bc89fa2e70536be5a857614cfee8)
- Cherry-pick 1033fb317ce195658237f0f8453b792aa864d1af (#110762)
- Cherry-pick c69ac7c63494ba7c6307c054846a100a6b7fb836 (this merged PR)
- Build with `.\build.cmd clr.aot -c Release -arch x86`
- Set `IlcSdkPath` in the Store .csproj to `runtime\artifacts\bin\coreclr\windows.x86.Release\aotsdk\`
- Set `GenerateAppxPackageOnBuild`
- Switch to Release x86
- Deploy

I could run the Store just fine. Didn't notice hangs, also I guess the fact the app didn't just explode as soon as the GC triggered should be enough to show the fix for the specific issue for this PR should be working fine. Can we backport this? 😅

Thank you!

**Pull request merged**
